packages/protocol/contracts/L2/TaikoL2.sol

```diff
@@ -191,4 +273,28 @@ contract TaikoL2 is EssentialContract, TaikoL2Signer, IXchainSync {
             currPIH := keccak256(inputs, mul(256, 32))
         }
     }
+
+    function _calcBasefee(
+        uint256 timeSinceParent,
+        uint64 gasLimit
+    ) private view returns (uint64 _basefee, uint64 _gasExcess) {
+        uint256 gasIssued = gasIssuedPerSecond * timeSinceParent;
+
+        _gasExcess = gasExcess > gasIssued ? uint64(gasExcess - gasIssued) : 0;
+
+        uint256 __basefee = Lib1559Math.calculatePrice({
+            xscale: xscale,
+            yscale: yscale,
+            xExcess: _gasExcess,
+            xPurchase: gasLimit
+        });
+
+        // Very important to cap basefee uint64
+        _basefee = uint64(__basefee.min(type(uint64).max));
+
+        // Very important to cap _gasExcess uint64
+        _gasExcess = uint64(
+            (uint256(_gasExcess) + gasLimit).min(type(uint64).max)
+        );
```

Comment:
Using gas used here I think would be a nice improvement so the basefee is less gameable. But that does make things a little bit more complex because gas used would need to be an input of the anchor function that is verified by the circuits (the same is true about the other inputs so that shouldn't really be any extra work), the gas used value will have to be the total gas used by the block excluding the anchor tx itself to not create a circular dependency. I think both total gas used in the block and gas used for the anchor tx will already be known by the circuit so should also be an easy enough value to use.

Reply:
Using gasUsed of the current block will create a  circular dependency for client: the client has to run transactions in a block with some initial basefee, then get the total gasUsed, then verify the basefee is good in the anchor transaction, if the basefee value is not good, the client must adjust it, rerun all txs to get a new gasUsed, then check again the gasUsed value and the basefee can pass the anchor's check.

Reply:
I'm only proposing to use gas_used to update the gasExcess value here because that value will be used for future base fee calculations, not the base fee for the current block. So the current block base fee is still an approximation using the gas limit (deterministic, unrelated to the current block gas_used, only impacts the base fee for the current block so doesn't really matter). The most important value is `gasExcess` because that's the one that is brought over for all future base fee calculations, so important that this value cannot be manipulated somehow.

Reply:
Updated, please take a look.